16566 update eslint rules

rianfowler opened the pull request:
- includes all of the config and script updates needed to run js linting from the base of the project
- consolidates eslint / ignore into one- this seemed preferable to having `lerna` run a duplicated linting script in each package 

```
packages/documentation/services/github-api.js:
packages/documentation/src/components/indexPage.js:
packages/documentation/src/components/propstable.js:
packages/documentation/src/components/seo.js:
packages/documentation/src/components/sidebar/Sidebar.js:
packages/documentation/src/components/sidebar/SidebarItems.jsx:
packages/documentation/src/layouts/external-layout.js:
packages/documentation/src/layouts/layout.js:
packages/documentation/src/layouts/module-components.js:
packages/documentation/src/pages/404.js:
packages/documentation/src/sidebar.js:
packages/formation-react/scripts/build.js:
packages/formation-react/src/components/AcceptTermsPrompt/AcceptTermsPrompt.jsx:
packages/formation-react/src/components/AcceptTermsPrompt/AcceptTermsPrompt.unit.spec.jsx:
packages/formation-react/src/components/AdditionalInfo/AdditionalInfo.jsx:
packages/formation-react/src/components/AdditionalInfo/AdditionalInfo.unit.spec.jsx:
packages/formation-react/src/components/AlertBox/AlertBox.jsx:
packages/formation-react/src/components/AlertBox/AlertBox.unit.spec.jsx:
packages/formation-react/src/components/Breadcrumbs/Breadcrumbs.jsx:
packages/formation-react/src/components/Breadcrumbs/Breadcrumbs.unit.spec.jsx:
packages/formation-react/src/components/CollapsiblePanel/CollapsiblePanel.jsx:
packages/formation-react/src/components/CollapsiblePanel/CollapsiblePanel.unit.spec.jsx:
packages/formation-react/src/components/DropDownPanel/DropDownPanel.jsx:
packages/formation-react/src/components/DropDownPanel/DropDownPanel.unit.spec.jsx:
packages/formation-react/src/components/ErrorableCheckbox/ErrorableCheckbox.jsx:
packages/formation-react/src/components/ErrorableCheckbox/ErrorableCheckbox.unit.spec.jsx:
packages/formation-react/src/components/ErrorableCheckboxGroup/ErrorableCheckboxGroup.jsx:
packages/formation-react/src/components/ErrorableCheckboxGroup/ErrorableCheckboxGroup.unit.spec.jsx:
packages/formation-react/src/components/ErrorableDate/ErrorableDate.jsx:
packages/formation-react/src/components/ErrorableDate/ErrorableDate.unit.spec.jsx:
packages/formation-react/src/components/ErrorableFileInput/ErrorableFileInput.jsx:
packages/formation-react/src/components/ErrorableFileInput/ErrorableFileInput.unit.spec.jsx:
packages/formation-react/src/components/ErrorableMonthYear/ErrorableMonthYear.jsx:
packages/formation-react/src/components/ErrorableMonthYear/ErrorableMonthYear.unit.spec.jsx:
packages/formation-react/src/components/ErrorableNumberInput/ErrorableNumberInput.jsx:
packages/formation-react/src/components/ErrorableNumberInput/ErrorableNumberInput.unit.spec.jsx:
packages/formation-react/src/components/ErrorableRadioButtons/ErrorableRadioButtons.jsx:
packages/formation-react/src/components/ErrorableRadioButtons/ErrorableRadioButtons.unit.spec.jsx:
packages/formation-react/src/components/ErrorableSelect/ErrorableSelect.jsx:
packages/formation-react/src/components/ErrorableSelect/ErrorableSelect.unit.spec.jsx:
packages/formation-react/src/components/ErrorableTextArea/ErrorableTextArea.jsx:
packages/formation-react/src/components/ErrorableTextArea/ErrorableTextArea.unit.spec.jsx:
packages/formation-react/src/components/ErrorableTextInput/ErrorableTextInput.jsx:
packages/formation-react/src/components/ErrorableTextInput/ErrorableTextInput.unit.spec.jsx:
packages/formation-react/src/components/ExpandingGroup/ExpandingGroup.jsx:
packages/formation-react/src/components/ExpandingGroup/ExpandingGroup.unit.spec.jsx:
packages/formation-react/src/components/HelpMenu/HelpMenu.jsx:
packages/formation-react/src/components/HelpMenu/HelpMenu.unit.spec.jsx:
packages/formation-react/src/components/IconBase/IconBase.jsx:
packages/formation-react/src/components/IconHelp/IconHelp.jsx:
packages/formation-react/src/components/IconSearch/IconSearch.jsx:
packages/formation-react/src/components/IconUser/IconUser.jsx:
packages/formation-react/src/components/LoadingIndicator/LoadingIndicator.jsx:
packages/formation-react/src/components/LoadingIndicator/LoadingIndicator.unit.spec.jsx:
packages/formation-react/src/components/MegaMenu/arrow-right-blue.jsx:
packages/formation-react/src/components/MegaMenu/Column.jsx:
packages/formation-react/src/components/MegaMenu/MegaMenu.jsx:
packages/formation-react/src/components/MegaMenu/MegaMenu.unit.spec.jsx:
packages/formation-react/src/components/MegaMenu/MenuSection.jsx:
packages/formation-react/src/components/MegaMenu/SubMenu.jsx:
packages/formation-react/src/components/modal/Modal.jsx:
packages/formation-react/src/components/modal/Modal.unit.spec.jsx:
packages/formation-react/src/components/OMBInfo/OMBInfo.jsx:
packages/formation-react/src/components/OMBInfo/OMBInfo.unit.spec.jsx:
packages/formation-react/src/components/Pagination/Pagination.unit.spec.jsx:
packages/formation-react/src/components/PrivacyAgreement/PrivacyAgreement.jsx:
packages/formation-react/src/components/PrivacyAgreement/PrivacyAgreement.unit.spec.jsx:
packages/formation-react/src/components/ProgressBar/ProgressBar.jsx:
packages/formation-react/src/components/ProgressBar/ProgressBar.unit.spec.jsx:
packages/formation-react/src/components/ProgressButton/ProgressButton.jsx:
packages/formation-react/src/components/ProgressButton/ProgressButton.unit.spec.jsx:
packages/formation-react/src/components/SearchMenu/SearchMenu.jsx:
packages/formation-react/src/components/SearchMenu/SearchMenu.unit.spec.jsx:
packages/formation-react/src/components/SegmentedProgressBar/SegmentedProgressBar.jsx:
packages/formation-react/src/components/SegmentedProgressBar/SegmentedProgressBar.unit.spec.jsx:
packages/formation-react/src/components/SortableTable/SortableTable.jsx:
packages/formation-react/src/components/SortableTable/SortableTable.unit.spec.jsx:
packages/formation-react/src/components/SystemDownView/SystemDownView.jsx:
packages/formation-react/src/components/SystemDownView/SystemDownView.unit.spec.jsx:
packages/formation-react/src/helpers/fields.js:
packages/formation-react/src/helpers/options-for-select.js:
packages/formation-react/src/helpers/test-helpers.js:
packages/formation-react/src/helpers/validations.js:
packages/formation-react/SubMenu.js:
packages/formation-react/SystemDownView.js:
packages/formation-react/test-helpers.js:
packages/formation-react/testing/tests.js:
packages/formation-react/validations.js:
packages/formation/scripts/build.js:
packages/vagov-eslint/custom-eslint/va-enzyme/index.js:
prebuild.js:
```

rianfowler commented on .eslintignore:
```diff
@@ -1,10 +1,7 @@
-fractalEntry.js
-vets-sketchapp/*
-/src/utils/*
+**/build/*
```
Not sure how we'd want to handle this if our linting was in a separate package- i guess we could try to allow each package pass in its own `eslintignore` but I don't know if we can merge that with another set of rules

rianfowler commented on .eslintrc:
```diff
@@ -1,104 +1,137 @@
 {
   // All rules should be disabled or they should produce errors. No warnings.
-  'extends': 'airbnb',
```
copied this directly from `vets-website` with no changes

rianfowler commented on .eslintignore:
```diff
@@ -1,10 +1,7 @@
+**/node_modules/*
+**/.cache/
+**/*.config.*
+**/*.conf.*
```
not sure if we want to lint our webpack / karma configs.

jbalboni replied:
Probably not

jbalboni commented on package.json:
```diff
@@ -11,8 +11,12 @@
-    "lint": "lerna run lint",
-    "lint:js": "lerna run lint",
+    "lint": "npm run lint:js",
+    "lint:changed": "npm run lint:js:changed && npm run lint:sass",
```
I don't think `lint:sass` exists.

jbalboni commented on package.json:
```diff
@@ -11,8 +11,12 @@
     "bootstrap": "lerna bootstrap",
-    "lint": "lerna run lint",
-    "lint:js": "lerna run lint",
+    "lint": "npm run lint:js",
```
I'm interested in what others thing about how to set this up. We could let each package control its linting command, which is sort of how it works now. Or we can run the linter commands from the root here. I sort of prefer the former, but it's a little more duplication of commands.

rianfowler replied:
That's how I approached it originally but walked back to this after having to duplicate all of the eslint plugins into each of the package.jsons. I'm fine with either way- I sort of see linting as a repo-level concern if we're just going to use the same rules across the project.

rianfowler commented on packages/vagov-eslint/package.json:
```diff
@@ -0,0 +1,24 @@
+    "dependencies": {
+        "eslint": "4.4.1",
+        "eslint-config-airbnb": "15.1",
+        "eslint-config-prettier": "^4.1.0",
```
Adds the eslint plugins to the dependency graph.

rianfowler commented on packages/vagov-eslint/package.json:
```diff
@@ -0,0 +1,24 @@
+    "description": "Shared JS linting package",
+    "version": "1.0.0",
+    "bin": {
+        "vagov-eslint": "./bin/vagov-eslint"
```
Exports our existing `package.json` scripts as a bin script- this allows them to be executed like `npx vagov-eslint lint:js`.

rianfowler commented on packages/vagov-eslint/bin/vagov-eslint:
```diff
@@ -0,0 +1,21 @@
+    shell.exec("LIST=`git diff-index --name-only HEAD | grep \"\\.js[x]\\{0,1\\}$\"`; if [ \"$LIST\" ]; then eslint --fix --quiet $LIST; fi")
+    break;
+  case 'lint:js:fix':
+    shell.exec("eslint --format compact --fix --quiet --ext .js --ext .jsx .");
```
pretty much verbatim the existing `package.json` linting scripts

rianfowler commented on packages/vagov-eslint/package.json:
```diff
@@ -0,0 +1,24 @@
+        "prettier": "^1.16.4"
+    },
+    "scripts": {
+        "postinstall" : "cp .eslintrc $INIT_CWD && cp .eslintignore $INIT_CWD"
```
`postinstall` is executed relative to the node module. `$INIT_CWD` is a reference to the directory where the install script is called. 

This copies the eslint configs making them available to the `npx` commands.

rianfowler commented on packages/formation/package.json:
```diff
@@ -28,6 +29,9 @@
     "moment": "^2.23.0"
   },
+  "devDependencies": {
+    "vagov-eslint": "file:../vagov-eslint"
```
add the eslint package as a dev dependency

rianfowler commented on packages/formation/package.json:
```diff
@@ -18,7 +18,8 @@
   "scripts": {
-    "build": "node ./scripts/build.js"
+    "build": "node ./scripts/build.js",
+    "lint": "npx vagov-eslint lint:js"
```
add the `lint` script which calls the `npx` command

jbalboni replied:
Do you need `npx` in a task in here? I think npm/yarn will work without it (similar to the tasks that run lerna).

rianfowler commented on package.json:
```diff
@@ -10,9 +10,7 @@
-    "bootstrap": "lerna bootstrap",
-    "lint": "lerna run lint",
-    "lint:js": "lerna run lint",
+    "bootstrap": "lerna bootstrap --npm-client=yarn --force-local",
```
This is where this implementation fails. I was able to get the linting to work with the steps above when doing an `yarn install` in each `package` but lerna has some limitations: 
- the bootstrap command is trying to reconcile cross package dependencies 
- this process isn't a straight `yarn install` in each package. 
  - it doesn't install locally referenced files- it only installs external dependencies 
  - `$INIT_CWD` is not localized to each package - it references the monorepo root (this this expected but it would be nice if it had some sort of replacement)

I tried these flags but I don't understand the difference between `bootstrap` and `install` well enough to know if this approach is feasible. Will spend a little more time on this but if it doesn't work I'll just fallback to either having the linting at the monorepo level or pulling everything into each one of the packages.

rianfowler replied:
sidenote: i'm going to leave this `--npm-client` flag so we'll probably get some `yarn.lock` updates in this. i can PR this separately if you want.

rianfowler replied:
I see that now- I assumed it wasn't set since I got a few `yarn.lock` updates.

jbalboni commented on packages/formation-react/package.json:
```diff
@@ -31,7 +31,9 @@
     "prop-types": "^15.6.2",
     "react-transition-group": "1",
-    "window-or-global": "^1.0.1"
+    "shelljs": "0.8.3",
```
Two of these look like dev dependencies

rianfowler replied:
fixed

jbalboni commented:
@rianfowler Would it be easier if we made the vagov-eslint "package" in here a config page, like ESLint outlines here: https://eslint.org/docs/developer-guide/shareable-configs?

It seems like that might cut down on some of the scripting required and allow us to share the important part of the config (though not the commands, I guess).

rianfowler commented on package.json:
```diff
@@ -10,11 +10,15 @@
+    "bootstrap": "lerna bootstrap --npm-client=yarn --force-local",
     "build": "lerna run build",
     "deploy": "lerna run deploy --scope=vagov-documentation",
+    "lint": "npx vagov-eslint lint:js",
```
Using `npm` here errors out. I'm not sure what `npm vagov-eslint` would target- my understanding is that `npx` runs scripts in `node_modules/bin` and that's there's not a way to run a `node_module`'s `package.json` scripts with an `npm` command.

rianfowler commented on package.json:
```diff
@@ -71,6 +70,7 @@
     "svg-url-loader": "^2.3.0",
     "url-loader": "^0.6.2",
     "uswds": "1.4.2",
+    "vagov-eslint": "file:./packages/vagov-eslint",
```
If we decide to publish this package, I think we can move this import into each one of the packages and `lerna` will install it correctly.

jbalboni commented:
@rianfowler There are a bunch of single quote to double quote changes in here, I think maybe the Prettier rules are being picked up correctly?

rianfowler commented:
@jbalboni fixed the prettier issue- I didn't know that it was more than an `eslint` plugin with it's own config. I added the config to the linting package and updated the postinstall to copy it over.

rianfowler commented:
Green light.
